man/man1/cdf2grid.1
.TH cdf2grid 1  "USER COMMANDS"
.SH NAME
cdf2grid \- convert grid NetCDF file into grid CLM or raw file
.SH SYNOPSIS
.B cdf2grid
[\-var name] [\-{float|double}] [\-scale s] [\-raw] [\-json]
.I netcdffile gridfile
.SH DESCRIPTION
Program converts grid file written in NetCDF format to grid file in raw or CLM format.
.SH OPTIONS
.TP
\-var name
Variable name in NetCDF file. If not specified the first variable with 2 dimensions is chosen.
.TP
\-scale factor
Scaling factor for CLM grid file. Default is 0.01 for short values in CLM grid file, 1 for float or double values. If factor differs from 1 and -float or -double option is set a warning is printed and scaling factor reset to 1.
.TP
\-{float|double}
Write float or double values in CLM grid file, default is short.
.TP
\-raw
Output format is raw without header, default is CLM format.
.TP
\-json
An additional JSON metafile with suffix \fI.json\fP is written.
.TP
.I netcdffile
Filename of grid NetCDF file.
.TP
.I gridfile
Filename of grid data file written.
.SH EXAMPLES
.TP
Create grid binary raw file and JSON metafile from grid NetCDF file:
.B cdf2grid
-raw -json grid.nc grid.bin
.PP
.SH EXIT STATUS
.B cdf2grid
returns a zero exit status if the CLM file is created.
Non zero is returned in case of failure.

.SH AUTHORS

For authors and contributors see AUTHORS file

.SH COPYRIGHT

(C) Potsdam Institute for Climate Impact Research (PIK), see COPYRIGHT file

.SH SEE ALSO
cdf2coord(1), cdf2clm(1), clm2cdf(1), bin2cdf(1), country2cdf(1), regridlpj(1), lpjml(1), txt2clm(1), cru2clm(1), txt2grid(1), clm(5)


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project follows a [GNU standard version numbering](https://www.gnu.org/prep/standards/html_node/Releases.html#index-version-numbers_002c-for-releases)
of `major.minor.patch` with
- `major` for substantial changes to code functionality,
- `minor` for additions of modules and functionality, and
- `patch` for technical changes and bugfixes without extending the functionality of the code.

- if applicable, categories to include in the CHANGELOG.md are:
  - Added
  - Changed
  - Deprecated
  - Removed
  - Fixed
  - Security

## [Unreleased]


### Contributors

- author: Werner von Bloh (bloh@pik-potsdam.de)

### Changed

- Option `-check` of `configure.sh` enables now run-time checks of memory leaks and memory access out of bounds and udefined variables for gcc and icx compiler. Optimization is not disabled. Use options `-debug -check` to disable.
- Datatype `List` used in `fprintfiles.c`.
- All allocated memory is now freed in `getcountry.c`, `lpjfiles.c`, `lpjcheck.c`, `mathclm.c`, `reservoir2cdf.c`, `printglobal.c` and `printharvest.c`.

### Fixed

- Check for missing values for integer input corrected in `readclimate_netcdf.c`.
- Memory leak fixed in `filesexist.c`.
- Calculation of global area fixed in `printglobal.c`.
- Check for null attribute pointer added in `fprintjson.c`.
- Uninitialized grid name initialized in `mathclm.c`.
- Call to `snprintf()` replaced by `getsprintf()` in `reservoir2cdf.c`.
- Datatype for index and writing the header corrected in `lpjcat.c`.
- Functions for qsort(), bisect(), and leftmostzero() changed to avoid run-time errors with `-check` option.
- Pointer set to zero  if `k_est`is zero in `initmanage.c`.
- Missing argument added in opening aquifer file for IMAGE.
- Scaling of coordinates fixed in `joingrid.c`.


## [5.9.22] - 2025-03-14

### Contributors

- author: Werner von Bloh (bloh@pik-potsdam.de), Hester Biemans (hester.biemans@wur.nl)
- code review: Susanne Rolinski (rolinski@pik-potsdam.de), Marie Hemmen (hemmen@pik-potsdam.de)

### Changed

- `input_netcdf.cjson` changed to the default dataset of `input.cjson` but in NetCDF format.
- Code changed to compile under Windows OS.
- New function `getsprintf()` added to allocate and print formatted output into string. Function replaces call to `snprintf()` function.
- Default settings for `GIT_HASH` and `GIT_REPO` added.
- Duplicate filenames removed from list of input/output filenames in utility `lpjfiles`.

### Added

- Option `-int` added to utility `cdf2clm`.
- Option `-latlon` added to utility `cdf2coord` in order to change the order of the CLM grid file.
- Utility `cdf2reservoir` added to convert NetCDF reservoir file into CLM file.
- Utility `reservoir2cdf` added to convert CLM reservoir file into a NetCDF file using the soil code NetCDF file.
- Macro `NETCDF_INPUT` added in `lpjml_config.cjson` to enable NetCDF input.
- Reservoir, irrigation neighbor and drainage data can now be in NetCDF format:
```java
"drainage" :           { "fmt" : "cdf", "var" : "index", "name" : "cru_netcdf/drainage.nc"},
"neighb_irrig" :       { "fmt" : "cdf", "var" : "index", "name" : "cru_netcdf/neighb_irrig.nc"},
"river" :              { "fmt" : "cdf", "var" : "riverlen", "name" : "cru_netcdf/drainage.nc"},
"reservoir" :          { "fmt" : "cdf", "var" : "year", "name" : "cru_netcdf/reservoir.nc"},
"capacity_reservoir" : { "fmt" : "cdf", "var" : "capacity", "name" : "cru_netcdf/reservoir.nc"},
"area_reservoir" :     { "fmt" : "cdf", "var" : "area", "name" : "cru_netcdf/reservoir.nc"},
"inst_cap_reservoir" : { "fmt" : "cdf", "var" : "inst_cap", "name" : "cru_netcdf/reservoir.nc"},
"height_reservoir" :   { "fmt" : "cdf", "var" : "height", "name" : "cru_netcdf/reservoir.nc"},
"purpose_reservoir" :  { "fmt" : "cdf", "var" : "purpose", "name" : "cru_netcdf/reservoir.nc"},
```

### Fixed

- River length is now correctly read in `initdrain.c` for data in NetCDF format.
- Man page of `cvrtclm` corrected.
- Missing check for open NetCDF file added to `openclimate.c`.


## [5.9.21] - 2025-03-13

### Contributors

- author: Werner von Bloh (bloh@pik-potsdam.de)
- code review: Jannes Breier (breier@pik-potsdam.de), Sebastian Ostberg (ostberg@pik-potsdam.de)

### Added

- `"coupled_host"` and `"coupled_port"` added to LPJmL configuration to specify where the coupled model is running and which port is used for communication.


## [5.9.20] - 2025-03-13

### Contributors

- author: Werner von Bloh (bloh@pik-potsdam.de)
- code review: Sebastian Ostberg (ostberg@pik-potsdam.de), Jens Heinke (heinke@pik-potsdam.de)

### Added

- Check for correct number of time steps added in `openclimate_netcdf.c`.
- Option `timestep` added to `setclm`.

### Changed

- Function `freegrid()` is calling `freecell()` to avoid redundant code.
- References in man pages updated.
- `-v` option of `cdf2clm` prints leap days setting for daily time step.
- Missing file `CHANGELOG.md` added to tar and zip file.
- Longitude and latitude boundaries excluded as variables in `cdf2grid`.
- Ids for sowing date and crop PHU input added to `couplerpar.h`.
- Length of GIT repository output limited in `copyright.c`.

### Fixed

- Memory leaks in utility `lpjprint` closed.
- Handling of NaN as missing value corrected in reading NetCDF files.
- Doubled initialization of standtype array corrected in `lpj_clinber4.c`.
- Option `type` corrected in `setclm`.
- Handling of different endianness corrected in `setclm`.
- If version is changed with `setclm` only version is updated in the clm file.
- Check that CO2 data is coming completely from socket corrected in `readco2.c`.
- Number of years are now calculated correctly from number of days in case of leap days if number of leap days reaches 365.
- Check for identical filenames corrected in `cutclm.c`.
- Missing `else` added in `readclimate()`.



src/utils/cdf2grid.c
/**************************************************************************************/
/**                                                                                \n**/
/**                c  d  f  2  g  r  i  d  .  c                                    \n**/
/**                                                                                \n**/
/**     Converts NetCDF data grid file into LPJ grid file in raw or CLM format     \n**/
/**                                                                                \n**/
/** (C) Potsdam Institute for Climate Impact Research (PIK), see COPYRIGHT file    \n**/
/** authors, and contributors see AUTHORS file                                     \n**/
/** This file is part of LPJmL and licensed under GNU AGPL Version 3               \n**/
/** or later. See LICENSE file or go to http://www.gnu.org/licenses/               \n**/
/** Contact: https://github.com/PIK-LPJmL/LPJmL                                    \n**/
/**                                                                                \n**/
/**************************************************************************************/

#include "lpj.h"

#define USAGE  "Usage: %s [-var name] [-{float|double}] [-scale s] [-raw] [-json] netcdffile gridfile\n"

#ifdef USE_NETCDF
#include <netcdf.h>

#define error(rc) if(rc){ fprintf(stderr,"ERROR421: Cannot read '%s': %s.\n",argv[iarg],nc_strerror(rc)); return EXIT_FAILURE;}

typedef struct
{
  int index;
  int ilon,ilat;
} Data;

static int cmp(const Data *a,const Data *b)
{
  return a->index-b->index;
} /* of 'cmp' */
#endif

int main(int argc,char **argv)
{
#ifdef USE_NETCDF
  int rc,ncid,var_id,dimids[2],i,j,nvars,lon_id,lat_id,ndims,x,y,iarg,len,n_attr=0;
  double *lat,*lon;
  float scalar=0.0;
  size_t lat_len,lon_len;
  int missing_value;
  Bool israw,isjson,scalar_set;
  char name[NC_MAX_NAME+1],*endptr;
  Header header;
  Intcoord coord;
  struct
  {
    double lon,lat;
  } coord_d;
  struct
  {
    float lon,lat;
  } coord_f;
  Netcdf_config netcdf_config;
  char *var;
  char *out_json,*arglist;
  char *source=NULL,*history=NULL,*title=NULL;
  FILE *out;
  Data *data;
  int *index;
  Attr *attrs=NULL;
  var=NULL;
  header.datatype=LPJ_SHORT;
  header.scalar=0.01;
  israw=isjson=scalar_set=FALSE;
  initsetting_netcdf(&netcdf_config);
  for(iarg=1;iarg<argc;iarg++)
    if(argv[iarg][0]=='-')
    {
      if(!strcmp(argv[iarg],"-var"))
      {
        if(argc==iarg+1)
        {
          fprintf(stderr,"Missing argument after option '-var'.\n"
                 USAGE,argv[0]);
          return EXIT_FAILURE;
        }
        var=argv[++iarg];
      }
      else if(!strcmp(argv[iarg],"-float"))
      {
        header.datatype=LPJ_FLOAT;
        header.scalar=1;
      }
      else if(!strcmp(argv[iarg],"-double"))
      {
        header.datatype=LPJ_DOUBLE;
        header.scalar=1;
      }
      else if(!strcmp(argv[iarg],"-raw"))
      {
        israw=TRUE;
      }
      else if(!strcmp(argv[iarg],"-json"))
      {
        isjson=TRUE;
      }
      else if(!strcmp(argv[iarg],"-scale"))
      {
        if(argc==iarg+1)
        {
          fprintf(stderr,"Missing argument after option '-scale'.\n"
                 USAGE,argv[0]);
          return EXIT_FAILURE;
        }
        scalar=header.scalar=(float)strtod(argv[++iarg],&endptr);
        if(*endptr!='\0')
        {
          fprintf(stderr,"Invalid number '%s' for scale.\n",argv[iarg]);
          return EXIT_FAILURE;
        }
        if(header.scalar!=1)
          scalar_set=TRUE;
      }
      else
      {
        fprintf(stderr,"Invalid option '%s'.\n"
                USAGE,argv[iarg],argv[0]);
        return EXIT_FAILURE;
      }
    }
    else
      break;
  if(argc<iarg+2)
  {
    fprintf(stderr,"Missing argument(s).\n"
            USAGE,argv[0]);
    return EXIT_FAILURE;
  }
  if(header.datatype!=LPJ_SHORT && scalar_set)
  {
    fprintf(stderr,"Warning: Scaling set to %g but datatype is %s, scaling set to 1.\n",
            scalar,typenames[header.datatype]);

    header.scalar=1;
  }
  rc=nc_open(argv[iarg],NC_NOWRITE,&ncid);
  if(rc)
  {
    fprintf(stderr,"ERROR409: Cannot open '%s': %s.\n",
            argv[iarg],nc_strerror(rc));
    return EXIT_FAILURE;
  }
  if(var==NULL)
  {
    nc_inq_nvars(ncid,&nvars);
    for(j=0;j<nvars;j++)
    {
      nc_inq_varname(ncid,j,name);
      if(strcmp(name,netcdf_config.lon_bnds.name) && strcmp(name,netcdf_config.lat_bnds.name))
      {
        nc_inq_varndims(ncid,j,&ndims);
        if(ndims==2)
        {
          var_id=j;
          break;
        }
      }
    }
    if(j==nvars)
    {
      fprintf(stderr,"ERRO405: No variable with 2 dimensions found in '%s'.\n",argv[iarg]);
      nc_close(ncid);
      return EXIT_FAILURE;
    }
  }
  else if(nc_inq_varid(ncid,var,&var_id))
  {
    fprintf(stderr,"ERROR406: Cannot find variable '%s' in '%s'.\n",var,argv[iarg]);
    nc_close(ncid);
    return EXIT_FAILURE;
  }
  nc_inq_varndims(ncid,var_id,&ndims);
  if(ndims!=2)
  {
    fprintf(stderr,"ERROR408: Invalid number of dimensions %d in '%s', must be 2.\n",
            ndims,argv[iarg]);
    nc_close(ncid);
    return EXIT_FAILURE;
  }
  nc_inq_vardimid(ncid,var_id,dimids);
  nc_inq_dimname(ncid,dimids[1],name);
  rc=nc_inq_varid(ncid,name,&lon_id);
  if(rc)
  {
    fprintf(stderr,"ERROR410: Cannot read %s in '%s': %s.\n",
            name,argv[iarg],nc_strerror(rc));
    nc_close(ncid);
    return EXIT_FAILURE;
  }
  nc_inq_dimlen(ncid,dimids[1],&lon_len);
  lon=newvec(double,lon_len);
  if(lon==NULL)
  {
    printallocerr("lon");
    nc_close(ncid);
    return EXIT_FAILURE;
  }
  rc=nc_get_var_double(ncid,lon_id,lon);
  if(rc)
  {
    free(lon);
    fprintf(stderr,"ERROR410: Cannot read longitude in '%s': %s.\n",
            argv[iarg],nc_strerror(rc));
    nc_close(ncid);
    return EXIT_FAILURE;
  }
  nc_inq_dimname(ncid,dimids[0],name);
  rc=nc_inq_varid(ncid,name,&lat_id);
  if(rc)
  {
    fprintf(stderr,"ERROR410: Cannot read %s in '%s': %s.\n",
            name,argv[iarg],nc_strerror(rc));
    free(lon);
    nc_close(ncid);
    return EXIT_FAILURE;
  }
  nc_inq_dimlen(ncid,dimids[0],&lat_len);
  lat=newvec(double,lat_len);
  if(lat==NULL)
  {
    free(lon);
    printallocerr("lat");
    nc_close(ncid);
    return EXIT_FAILURE;
  }
  rc=nc_get_var_double(ncid,lat_id,lat);
  if(rc)
  {
    free(lon);
    free(lat);
    fprintf(stderr,"ERROR410: Cannot read latitude in '%s': %s.\n",
            argv[iarg],nc_strerror(rc));
    nc_close(ncid);
    return EXIT_FAILURE;
  }

  rc=nc_get_att_int(ncid,var_id,"missing_value",&missing_value);
  if(rc)
    rc=nc_get_att_int(ncid,var_id,"_FillValue",&missing_value);
  if(rc)
  {
    fprintf(stderr,"WARNING402: Cannot read missing or fill value in '%s': %s, set to %d.\n",
            argv[iarg],nc_strerror(rc),netcdf_config.missing_value.i);
    missing_value=netcdf_config.missing_value.i;
  }
  header.cellsize_lon=(lon[lon_len-1]-lon[0])/(lon_len-1);
  header.cellsize_lat=(float)fabs((lat[lat_len-1]-lat[0])/(lat_len-1));
  index=newvec(int,lat_len*lon_len);
  if(index==NULL)
  {
    free(lon);
    free(lat);
    printallocerr("index");
    nc_close(ncid);
    return EXIT_FAILURE;
  }
  rc=nc_get_var_int(ncid,var_id,index);
  error(rc);
  if(isjson)
  {
    history=getattr_netcdf(ncid,NC_GLOBAL,"history");
    source=getattr_netcdf(ncid,NC_GLOBAL,"source");
    title=getattr_netcdf(ncid,NC_GLOBAL,"title");
    if(nc_inq_natts(ncid,&len))
      n_attr=0;
    else
    {
      attrs=newvec(Attr,len);
      if(attrs==NULL)
      {
        printallocerr("attrs");
        free(index);
        free(lon);
        free(lat);
        nc_close(ncid);
        return EXIT_FAILURE;
      }
      n_attr=0;
      for(i=0;i<len;i++)
      {
        if(!nc_inq_attname(ncid,NC_GLOBAL,i,name))
        {
          if(strcmp(name,"history") && strcmp(name,"source") && strcmp(name,"title"))
          {
            attrs[n_attr].value=getattr_netcdf(ncid,NC_GLOBAL,name);
            if(attrs[n_attr].value!=NULL)
              attrs[n_attr++].name=strdup(name);
          }
        }
      }
    }
  }
  nc_close(ncid);
  /* find number of valid cells in grid NetCDF file */
  header.ncell=0;
  for(i=0;i<lat_len*lon_len;i++)
    if(index[i]!=missing_value)
      header.ncell++;
  if(header.ncell==0)
  {
    fprintf(stderr,"No grid cells found in '%s'.\n",argv[iarg]);
    free(index);
    free(lon);
    free(lat);
    return EXIT_FAILURE;
  }
  data=newvec(Data,header.ncell);
  if(data==NULL)
  {
    free(index);
    free(lon);
    free(lat);
    printallocerr("data");
    return EXIT_FAILURE;
  }
  header.ncell=0;
  for(y=0;y<lat_len;y++)
    for(x=0;x<lon_len;x++)
    {
      if(index[y*lon_len+x]!=missing_value)
      {
        /* valid cell found, store index and ilat,ilon */
        data[header.ncell].index=index[y*lon_len+x];
        data[header.ncell].ilon=x;
        data[header.ncell].ilat=y;
#ifdef DEBUG
        printf("%d %d %d\n",index[y*lon_len+x],x,y);
#endif
        header.ncell++;
      }
    }
  free(index);
  /* sort in ascending order of index */
  qsort(data,header.ncell,sizeof(Data),(int(*)(const void *,const void *))cmp);
  header.firstcell=data[0].index;
  for(i=1;i<header.ncell;i++)
    if(data[i].index!=i+header.firstcell)
    {
      fprintf(stderr,"Missing or double index in grid NetCDF file '%s' at cell %d.\n",
              argv[iarg],i);
      free(data);
      free(lon);
      free(lat);
      return EXIT_FAILURE;
    }
  header.nyear=1;
  header.nstep=1;
  header.timestep=1;
  header.firstyear=1901;
  header.nbands=2;
  header.order=CELLYEAR;
  out=fopen(argv[iarg+1],"wb");
  if(out==NULL)
  {
    fprintf(stderr,"Error creating '%s': %s.\n",argv[iarg+1],strerror(errno));
    free(data);
    free(lon);
    free(lat);
    return EXIT_FAILURE;
  }
  if(!israw)
    fwriteheader(out,&header,LPJGRID_HEADER,LPJGRID_VERSION);
  for(i=0;i<header.ncell;i++)
  {
    switch(header.datatype)
    {
      case LPJ_FLOAT:
        coord_f.lat=(float)lat[data[i].ilat];
        coord_f.lon=(float)lon[data[i].ilon];
        rc=fwrite(&coord_f,sizeof(coord_f),1,out);
        break;
      case LPJ_DOUBLE:
        coord_d.lat=lat[data[i].ilat];
        coord_d.lon=lon[data[i].ilon];
        rc=fwrite(&coord_d,sizeof(coord_d),1,out);
        break;
      default:
        coord.lat=(short)(lat[data[i].ilat]/header.scalar);
        coord.lon=(short)(lon[data[i].ilon]/header.scalar);
#ifdef DEBUG
        printf("%.3f %3f %d %d\n",lat[data[i].ilat],lon[data[i].ilon],coord.lat,coord.lon);
#endif
        rc=fwrite(&coord,sizeof(coord),1,out);
    } /* of switch() */
    if(rc!=1)
    {
      fprintf(stderr,"Error writing grid file '%s': %s.\n",
              argv[iarg+1],strerror(errno));
      fclose(out);
      free(data);
      free(lon);
      free(lat);
      return EXIT_FAILURE;
    }
  }
  fclose(out);
  free(data);
  free(lon);
  free(lat);
  printf("Number of cells: %d\n",header.ncell);
  if(isjson)
  {
    out_json=malloc(strlen(argv[iarg+1])+strlen(JSON_SUFFIX)+1);
    if(out_json==NULL)
    {
      printallocerr("filename");
      return EXIT_FAILURE;
    }
    strcat(strcpy(out_json,argv[iarg+1]),JSON_SUFFIX);
    arglist=catstrvec(argv,argc);
    out=fopen(out_json,"w");
    if(out==NULL)
    {
      printfcreateerr(out_json);
      return EXIT_FAILURE;
    }
    free(out_json);
    fprintjson(out,argv[iarg+1],title,source,history,arglist,&header,NULL,NULL,attrs,n_attr,"grid","degree",NULL,"cell coordinates",NULL,LPJ_SHORT,(israw) ? RAW : CLM,LPJGRID_HEADER,FALSE,LPJGRID_VERSION);
    free(arglist);
    free(title);
    free(source);
    free(history);
    freeattrs(attrs,n_attr);
    fclose(out);
  }
  return EXIT_SUCCESS;
#else
  fprintf(stderr,"ERROR401: NetCDF is not supported in this version of %s.\n",argv[0]);
  return EXIT_FAILURE;
#endif
} /* of 'main' */
